### PR TITLE
Add retry to rsync in cacheservice

### PR DESCRIPTION
**Pull request by mimi1vx**
https://progress.opensuse.org/issues/103771

### CODE REVIEW

File: lib/OpenQA/CacheService/Task/Sync.pm
Comment (okurz) [inline lib/OpenQA/CacheService/Task/Sync.pm:50]:
```suggestion
use constant RSYNC_RETRIES => $ENV{OPENQA_RSYNC_RETRIES} // 7;
use constant RSYNC_RETRY_PERIOD => $ENV{OPENQA_RSYNC_RETRY_PERIOD} // 3;
…
    for my $retry (1 .. RSYNC_RETRIES) {
        my $output = `@cmd`;
        my $status = $? >> 8;
        $job->note(output => "[info] [#$job_id] Calling try $retry: $cmd\n$output");
        last unless $status;
        sleep RSYNC_RETRY_PERIOD;
    }
```

**Review by okurz — CHANGES_REQUESTED** [review]

File: lib/OpenQA/CacheService/Task/Sync.pm
Comment (mimi1vx) [reply lib/OpenQA/CacheService/Task/Sync.pm:50]: 7 retries ?

Comment (okurz) [reply lib/OpenQA/CacheService/Task/Sync.pm:50]: Yes, I always default to 7 retries as a magic number :) I am ok to keep 3 though, then just do the rest of the code suggestions

Comment (mimi1vx) [reply lib/OpenQA/CacheService/Task/Sync.pm]: j, usage of ENV vars is way better than some random numbers in code

File: lib/OpenQA/CacheService/Task/Sync.pm
Comment (kraih) [inline lib/OpenQA/CacheService/Task/Sync.pm:48]: "Calling calling"?

File: t/25-cache-service.t
Comment (kraih) [inline t/25-cache-service.t:107]: This does not actually test that it tried more than once.

Comment (Martchus) [reply t/25-cache-service.t:107]: Yes, and it would be nice to have a test.

File: lib/OpenQA/CacheService/Task/Sync.pm
Comment (kraih) [inline lib/OpenQA/CacheService/Task/Sync.pm:48]: Is it intentional that the latest result overwrites the previous one? I think for other retry features we log all attempts.

Comment (okurz) [reply lib/OpenQA/CacheService/Task/Sync.pm:48]: I guess it's ok either way

**Review by okurz — APPROVED** [review]

File: lib/OpenQA/CacheService/Task/Sync.pm
Comment (mimi1vx) [reply lib/OpenQA/CacheService/Task/Sync.pm:48]: no ,isn't , thx for spotting.

**Review by okurz — APPROVED** [review]

**Review by Martchus — APPROVED** [review]

File: lib/OpenQA/CacheService/Task/Sync.pm
Comment (okurz) [inline lib/OpenQA/CacheService/Task/Sync.pm:51]: To cover this in tests without needles waiting I suggest to use https://metacpan.org/pod/Test::Mock::Time which we already use in os-autoinst. You just need to specify the package as dependency in openQA but it's already installed in the CI container due to the os-autoinst dependency anyway

Comment (Martchus) [reply lib/OpenQA/CacheService/Task/Sync.pm:51]: Or you simply set `OPENQA_RSYNC_RETRY_PERIOD=0` in the test environment. (Needs to be in a `BEGIN` block as it is a constant.)

**Review by Martchus — APPROVED** [review]

File: lib/OpenQA/CacheService/Task/Sync.pm
Comment (kraih) [inline lib/OpenQA/CacheService/Task/Sync.pm:49]: This still throws away previous attempts in `$full_output`.

File: lib/OpenQA/CacheService/Task/Sync.pm
Comment (mimi1vx) [reply lib/OpenQA/CacheService/Task/Sync.pm:51]: > Or you simply set `BEGIN { $ENV{OPENQA_RSYNC_RETRY_PERIOD} = 0 }` in the test environment. (Needs to be in a `BEGIN` block as it is a constant.)

tried, not working:( so time for Test::Mock::Time

Comment (Martchus) [reply lib/OpenQA/CacheService/Task/Sync.pm:51]: Strange, it should've been working - especially since there are already similar overrides in the same test file.

Comment (mimi1vx) [reply lib/OpenQA/CacheService/Task/Sync.pm:51]: and `Test::More::Time` broke *Subtest: Minion monitoring with InfluxDB* , so probably best option is mark `sleep` line as non-coverable :(

Comment (Martchus) [reply lib/OpenQA/CacheService/Task/Sync.pm]: So you would want to mark the line as non-coverable because you want to put in some extra condition to not run the code at all? I'm just asking because the issue we want to avoid is that tests take longer due to the sleep.

File: lib/OpenQA/CacheService/Task/Sync.pm
Comment (Martchus) [inline lib/OpenQA/CacheService/Task/Sync.pm:51]: Ok, so this line isn't executed anyways by tests. How do you come to the conclusion that my suggestion to set `OPENQA_RSYNC_RETRY_PERIOD` to 0 isn't working then? Of course if the line isn't executed anyways we don't have to worry much about slowing down tests although it would likely be nice to actually have a test which runs all of the code.

**Review by Martchus — DISMISSED** [review]

File: lib/OpenQA/CacheService/Task/Sync.pm
Comment (mimi1vx) [reply lib/OpenQA/CacheService/Task/Sync.pm:51]: https://github.com/os-autoinst/openQA/pull/4464/files#diff-9a2b1ade12881dfa535761e60a94118eabd51ecdad9a003037400c6cad1df61dR114-R117 this should execute whole for loop include `sleep` statement ...

**Review by okurz — CHANGES_REQUESTED** [review]
.

File: lib/OpenQA/CacheService/Task/Sync.pm
Comment (Martchus) [reply lib/OpenQA/CacheService/Task/Sync.pm:51]: Then we shouldn't need to mark the line as uncoverable statement (unless it is running in a way which doesn't allow us to track its coverage, e.g. in a forked process).

Comment (mimi1vx) [reply lib/OpenQA/CacheService/Task/Sync.pm:51]: @kraih how `sleep` behave in minion job?

Comment (mimi1vx) [reply lib/OpenQA/CacheService/Task/Sync.pm]: > Then we shouldn't need to mark the line as uncoverable statement (unless it is running in a way which doesn't allow us to track its coverage, e.g. in a forked process).

so question is why codecov consider this line not running.. whed I added debug lines before and after they show correctly in test run

Comment (Martchus) [reply lib/OpenQA/CacheService/Task/Sync.pm]: Oh, when it is within a Minion job that's a forked process and therefore the coverage collection doesn't work. We actually have the utility function `perform_minion_jobs` to avoid forking which might be applicable here as well.

Comment (kraih) [reply lib/OpenQA/CacheService/Task/Sync.pm]: What @Martchus said. It's a bit tricky, since depending on test strategy Minion jobs might run without forking.

**Comment by okurz** [comment]
https://app.circleci.com/pipelines/github/os-autoinst/openQA/8847/workflows/585504f3-5933-4802-a650-7c2e8b48f245/jobs/83520 shows many error messages that should be avoided

**Comment by mimi1vx** [comment]
> https://app.circleci.com/pipelines/github/os-autoinst/openQA/8847/workflows/585504f3-5933-4802-a650-7c2e8b48f245/jobs/83520 shows many error messages that should be avoided

`rsync: .....` on stderr is excepted 3x per test run , and subtest itself calls `test_sync` 4x

File: lib/OpenQA/CacheService/Task/Sync.pm
Comment (mimi1vx) [reply lib/OpenQA/CacheService/Task/Sync.pm]: no idea how to use `perform_minion_jobs` in this case

**Comment by okurz** [comment]
> > https://app.circleci.com/pipelines/github/os-autoinst/openQA/8847/workflows/585504f3-5933-4802-a650-7c2e8b48f245/jobs/83520 shows many error messages that should be avoided
> 
> `rsync: .....` on stderr is excepted 3x per test run , and subtest itself calls `test_sync` 4x

You can use Test::Output to catch the expected error output. We use that in other tests as well. Just grep for `Test::Output`

**Comment by mimi1vx** [comment]
> You can use Test::Output to catch the expected error output. We use that in other tests as well. Just grep for `Test::Output`

I think this can test if `rsync` output is excepted , but not suppress error messages

File: t/25-cache-service.t
Comment (perlpunk) [inline t/25-cache-service.t:121]: One more thing you could do here is to test the actual output against a regex instead of just testing that it is not empty.

File: lib/OpenQA/CacheService/Task/Sync.pm
Comment (perlpunk) [reply lib/OpenQA/CacheService/Task/Sync.pm]: I'm trying to add `perform_minion_jobs`, but this will take a while.

File: t/25-cache-service.t
Comment (mimi1vx) [reply t/25-cache-service.t:121]: isnt this in next line ?

**Comment by perlpunk** [comment]
> I think this can test if `rsync` output is excepted , but not suppress error messages

I played with `perform_minion_jobs`, and when using that, I can catch the output with `stderr_like`.
However, it is not that trivial to add `perform_minion_jobs` just to one place. We discussed it in the weekly unblock meeting today, that it should be added to the whole test.
I could try that and if I get it working, I'll send you a commit.

File: t/25-cache-service.t
Comment (perlpunk) [reply t/25-cache-service.t:121]: oh I'm stupid. Ok, but maybe actually add the `Try 1:` stuff to the regex to make sure it actually retries

**Comment by okurz** [comment]
I also suggested an alternative which is to have simple unit-tests for the not covered parts and mock out the lower level code that is executed from minions for the higher-level tasks

File: t/25-cache-service.t
Comment (perlpunk) [inline t/25-cache-service.t:145]: That line and the next `unregister` line below were leftovers from my testing, they are not needed.

File: t/25-cache-service.t
Comment (perlpunk) [inline t/25-cache-service.t:313]: I think this line has to be added again

Comment (kraih) [reply t/25-cache-service.t:313]: Since you're running the jobs without forking you might be fine now without it.

Comment (perlpunk) [reply t/25-cache-service.t:313]: I saw in the codecov report that some lines lost their coverage:
https://app.codecov.io/gh/os-autoinst/openQA/compare/4464/changes#D1L339
So I would add it back and see if that changes.
There are still forked processes in the test, right? just not the cacheservice workers?

File: t/25-cache-service.t
Comment (perlpunk) [reply t/25-cache-service.t:7]: I'm so used to always put a `local` before changing env vars... :)

File: t/25-cache-service.t
Comment (okurz) [reply t/25-cache-service.t:313]: Yes, please ensure that t/25-cache-service.t has 100% statement coverage. It does not make sense to have lines in tests that are never or not reliably executed – with explicitly marked uncoverable exceptions of course.

File: t/25-cache-service.t
Comment (Martchus) [inline t/25-cache-service.t:402]: All tests now start their own workers, right? So this comment could be removed as well.

**Review by okurz — CHANGES_REQUESTED** [review]
Looks like unfortunately now the test is flaky *and* has unhandled output on stdout, see https://app.circleci.com/pipelines/github/os-autoinst/openQA/8889/workflows/6f71d3ec-2990-441a-bcfc-6d7f4d43e301/jobs/83929

File: t/25-cache-service.t
Comment (perlpunk) [reply t/25-cache-service.t]: hm, this line and the `unregister` line is still in there, just that you removed the comment

**Comment by perlpunk** [comment]
I guess the additional output is from `$ENV{MOJO_LOG_LEVEL} = 'trace';`
Maybe there is a way to only set that locally? (I tried `local` in the subtest, but that didn't work)

**Comment by perlpunk** [comment]
The test failed for me once when I added the `stderr_like` but then I enclosed more commands in it and then the test passed. But that was probably the flaky behaviour that we see in CI now :(

**Comment by perlpunk** [comment]
I pushed a commit that captures the unhandled log messages, while still assuring that the test excepting a log message `ok(scalar @logs, 'Output should be present')` is passing.

I wasn't able to fix the flakyness. The rsync errors sometimes happen outside of the `stderr_like`, no matter where I let it start and end.

**Review by Martchus — DISMISSED** [review]

**Review by okurz — APPROVED** [review]
https://app.circleci.com/pipelines/github/os-autoinst/openQA/8910/workflows/260a33d7-fae4-4c4d-be2b-b889e6af4631/jobs/84134 looks good so it seems like you now just try to cover a bigger block with the stderr catching so I guess that's ok

**Review by kraih — DISMISSED** [review]

**Comment by perlpunk** [comment]
> https://app.circleci.com/pipelines/github/os-autoinst/openQA/8910/workflows/260a33d7-fae4-4c4d-be2b-b889e6af4631/jobs/84134 looks good so it seems like you now just try to cover a bigger block with the stderr catching so I guess that's ok

no, i got a failure with this code locally. 
i mentioned this above.

**Comment by okurz** [comment]
@perlpunk so what do you suggest we do? This PR is waiting for your review now

**Comment by kalikiana** [comment]
> @perlpunk so what do you suggest we do? This PR is waiting for your review now

There's two approvals and no open requests for changes. That should be enough

**Comment by perlpunk** [comment]
My review was explicitly requested.
But @okurz was the one who saw the flakyiness https://github.com/os-autoinst/openQA/pull/4464#pullrequestreview-866277438
But if @okurz still approves the PR then I guess it's ok for him.
I still would like to find out what's happening.
Since it's not happening that often we could
* accept this PR
* create a ticket to investigate the problem with the error output

**Review by okurz — CHANGES_REQUESTED** [review]
I might have wrongly assumed that mimi1vx would have addressed your comment so I assumed the sporadic issue would have been addressed. So better we fix it here

**Comment by kraih** [comment]
Guess i'll also invalidate my review then, since i'm a bit confused now. From the looks of it, the code seems reasonable and CI tests passed... 🤷

**Comment by mimi1vx** [comment]
on my instance tests were flaky also before change, but I'm not experienced with test suite enough to decide cause or outcome..

**Comment by okurz** [comment]
yes, it's correct that t/25-cache-service.t is unstable. This is why it's part of tools/unstable_tests.txt. It would be nice to fix the flakyness but here the focus was on changing behaviour in the log output. That's how I understood. @perlpunk can you help to state more explicitly how you currently reproduce failures and which ones in particular?

**Comment by perlpunk** [comment]
I was getting test failures where the rsync error messages were written outside of the `stderr_like` and the `stderr_like` failed because it didn't catch anything.
```
rsync: change_dir "/home/squamata/project//empty" failed: No such file or directory (2)
rsync: mkdir "/home/squamata/project/nowhewre/tests" failed: No such file or directory (2)
rsync error: error in file IO (code 11) at main.c(664) [Receiver=3.1.3]
...
```
It was happening like 1 / 4 times or so.
Now I ran it 25 times without a failure, although the code is the same.

So I guess I'll approve.

**Review by perlpunk — APPROVED** [review]

**Review by okurz — APPROVED** [review]

**Comment by perlpunk** [comment]
argh. I rebased to master now, and I got the failure at the 6th run. I think last time when I tested it I had it also rebased on master.

I rebase my branches on master as often as possible, because of things like this.

**Comment by Martchus** [comment]
Maybe the instability is unrelated then (e.g. would also occur on master before merging this PR).

**Comment by perlpunk** [comment]
for the record:
```
t/25-cache-service.t .. 17/? rsync: change_dir "/home/tina/openqa-devel/repos/openQA//empty" failed: No such file or directory (2)
rsync: mkdir "/home/tina/openqa-devel/repos/openQA/nowhewre/tests" failed: No such file or directory (2)
rsync error: error in file IO (code 11) at main.c(664) [Receiver=3.1.3]
rsync: change_dir "/home/tina/openqa-devel/repos/openQA//empty" failed: No such file or directory (2)
rsync: mkdir "/home/tina/openqa-devel/repos/openQA/nowhewre/tests" failed: No such file or directory (2)
rsync error: error in file IO (code 11) at main.c(664) [Receiver=3.1.3]
rsync: change_dir "/home/tina/openqa-devel/repos/openQA//empty" failed: No such file or directory (2)
rsync: mkdir "/home/tina/openqa-devel/repos/openQA/nowhewre/tests" failed: No such file or directory (2)
rsync error: error in file IO (code 11) at main.c(664) [Receiver=3.1.3]

    #   Failed test 'rsync error message'
    #   at t/25-cache-service.t line 128.
    # STDERR:
    # 
    # doesn't match:
    # (?^:rsync error:.*)
    # as expected
    # Looks like you failed 1 test of 44.
t/25-cache-service.t .. 22/? 
#   Failed test 'OpenQA::CacheService::Task::Sync'
#   at t/25-cache-service.t line 647.
t/25-cache-service.t .. 23/? # Looks like you failed 1 test of 23.
t/25-cache-service.t .. Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/23 subtests 

```

**Comment by perlpunk** [comment]
Ok, I tried to bisect and apparently it was a coincidence. So it seems to be really rare.

**Pull request merged**
